dispatcher/dispatcherV2.go

```diff
@@ -0,0 +1,250 @@
+		return
+	}
+	msg := &message{
+		ctx:         ctx,
+		chainID:     chainID,
+		msg:         msgProto,
+		peer:        peer,
+		msgType:     msgType,
+		isBroadcast: true,
+	}
+	queue := d.queueForMsg(msg)
+	select {
+	case queue <- msg:
+	default:
+		log.L().Warn("Broadcast queue is full.", zap.Any("msgType", msgType))
+	}
+
+	d.updateMetrics(msg, queue)
+}
+
+func (d *dispatcherV2) HandleTell(ctx context.Context, chainID uint32, peer peer.AddrInfo, msgProto proto.Message) {
+	msgType, err := goproto.GetTypeFromRPCMsg(msgProto)
+	if err != nil {
+		log.L().Warn("Unexpected message handled by HandleTell.", zap.Error(err))
+	}
+	msg := &message{
+		ctx:      ctx,
+		chainID:  chainID,
+		msg:      msgProto,
+		peerInfo: peer,
+		msgType:  msgType,
+	}
+	queue := d.queueForMsg(msg)
+	select {
+	case queue <- msg:
+	default:
+		log.L().Warn("Unicast queue is full.", zap.Any("msgType", msgType))
+	}
+
+	d.updateMetrics(msg, queue)
+}
+
+func (d *dispatcherV2) AddSubscriber(chainID uint32, subscriber Subscriber) {
+	d.subscribersMU.Lock()
+	d.subscribers[chainID] = subscriber
+	d.subscribersMU.Unlock()
+}
+
+func (d *dispatcherV2) updateMetrics(msg *message, queue chan *message) {
+	subscriber := d.subscriber(msg.chainID)
+	if subscriber != nil {
+		subscriber.ReportFullness(msg.ctx, msg.msgType, float32(len(queue))/float32(cap(queue)))
+	}
+}
+
+func (d *dispatcherV2) dispatchBroadcastMsg(message *message) {
+	if !d.IsReady() {
+		return
+	}
+	subscriber := d.subscriber(message.chainID)
+	if subscriber == nil {
+		log.L().Warn("chainID has not been registered in dispatcher.", zap.Uint32("chainID", message.chainID))
+		return
+	}
+	switch msg := message.msg.(type) {
+	case *iotextypes.ConsensusMessage:
+		if err := subscriber.HandleConsensusMsg(msg); err != nil {
+			log.L().Warn("Failed to handle consensus message.", zap.Error(err))
+		}
+	case *iotextypes.Action:
+		if err := subscriber.HandleAction(message.ctx, msg); err != nil {
+			requestMtc.WithLabelValues("AddAction", "false").Inc()
+			log.L().Warn("Handle action request error.", zap.Error(err))
+		}
+	case *iotextypes.Actions:
+		for i := range msg.Actions {
+			if err := subscriber.HandleAction(message.ctx, msg.Actions[i]); err != nil {
+				requestMtc.WithLabelValues("AddAction", "false").Inc()
+				log.L().Warn("Handle action request error.", zap.Error(err))
+			}
+		}
+	case *iotextypes.Block:
+		if err := subscriber.HandleBlock(message.ctx, message.peer, msg); err != nil {
+			log.L().Error("Fail to handle the block.", zap.Error(err))
+		}
+	case *iotextypes.NodeInfo:
+		if err := subscriber.HandleNodeInfo(message.ctx, message.peer, msg); err != nil {
+			log.L().Warn("Failed to handle node info message.", zap.Error(err))
+		}
+	default:
+		msgType, _ := goproto.GetTypeFromRPCMsg(message.msg)
+		log.L().Warn("Unexpected msgType handled by HandleBroadcast.", zap.Any("msgType", msgType))
+	}
+}
+
+func (d *dispatcherV2) dispatchUnicastMsg(message *message) {
```

Comment:
the logic actually same as `dispatchBroadcastMsg`, can combine into one func

Reply:
conbined